enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md

```diff
@@ -0,0 +1,149 @@
+---
+title: ETCD-Encryption-For-Separate-OAuth-APIs
+authors:
+  - "@p0lyn0mial"
+reviewers:
+  - "@sttts"
+approvers:
+  - "@derekwaynecarr"
+creation-date: 2020-03-17
+last-updated: 2020-03-17
+status: implementable
+see-also: https://github.com/openshift/enhancements/blob/master/enhancements/authentication/separate-oauth-resources.md, https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/
+replaces:
+superseded-by:
+---
+
+# ETCD Encryption For Separate OAuth APIs
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+## Summary
+
+The `encryption-config` used by OpenShift API server to encrypt/decrypt resources will be also used by the new `oauth-apiserver`.
+Initially `OAS-O` will be responsible to manage both servers. In the future releases `CAO` will take over the config and will manage its operand.
+
+## Motivation
+
+Starting from version `4.3` customers that want to have additional layer of data security can enable etcd encryption. Once enabled OpenShift API server encrypts among others `OAuth tokens`.
+In version `4.n` we decided to split `openshift-apiserver` and create a new server called `oauth-apiserver`.
+That means the resources previously managed by a different component now will be served by a brand new API server.
+That also means that `oauth-apiserver` needs to be given the encryption keys in order to decrypt previously encrypted resources, like the aforementioned tokens.
+
+
+Additionally, we would like to provide a smooth transition for our customers. 
+Without any manual interaction and with fully tested and working upgrade and downgrade paths. 
+This document describes how we are going to achieve that.
+
+### Goals
+1. Make it possible to run the OpenShift OAuth API Server on an encrypted cluster.
+
+### Non-Goals
+
+
+## Proposal
+
+In OpenShift encryption at rest leverages Kubernetes built-in mechanism. It is based on `EncryptionConfiguration` that controls how API data is encrypted in etcd.
+It holds an array of keys that are used to encrypt resources. 
```

Comment:
It holds one array of keys for each resource.

Reply:
actually the array will eventually be equal, won't they? But we still have these multiple arrays because that's how upstream does it.